### PR TITLE
FastCorrelativeScanMatcher outputs unique_ptr.

### PR DESCRIPTION
**Pull request by gaschler**
Match methods use unique_ptr to output an optional result rather than writing into multiple pointer arguments.

### CODE REVIEW

File: cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher_test.cc [hunk 1]
```diff
@@ -140,26 +140,22 @@ TEST_F(FastCorrelativeScanMatcherTest, CorrectPoseForMatch) {
...
-    float rotational_score = 0.f;
-    float low_resolution_score = 0.f;
-    EXPECT_TRUE(fast_correlative_scan_matcher->Match(
-        transform::Rigid3d::Identity(), transform::Rigid3d::Identity(),
-        CreateConstantData(point_cloud_), kMinScore, &score, &pose_estimate,
-        &rotational_score, &low_resolution_score));
-    EXPECT_LT(kMinScore, score);
-    EXPECT_LT(0.09f, rotational_score);
-    EXPECT_LT(0.14f, low_resolution_score);
+    std::unique_ptr<FastCorrelativeScanMatcher::Result> result =
```
Comment (wohe) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher_test.cc]: `const`?

File: cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc [hunk 2]
```diff
@@ -179,34 +179,36 @@ void ConstraintBuilder::ComputeConstraint(
...
   if (match_full_submap) {
-    if (submap_scan_matcher->fast_correlative_scan_matcher->MatchFullSubmap(
+    match_result =
+        submap_scan_matcher->fast_correlative_scan_matcher->MatchFullSubmap(
             global_node_pose.rotation(), global_submap_pose.rotation(),
-            *constant_data, options_.global_localization_min_score(), &score,
-            &pose_estimate, &rotational_score, &low_resolution_score)) {
-      CHECK_GT(score, options_.global_localization_min_score());
+            *constant_data, options_.global_localization_min_score());
+    if (match_result) {
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc]: Consistency: Be explicit and compare ` != nullptr`.

File: cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc [hunk 3]
```diff
@@ -179,34 +179,36 @@ void ConstraintBuilder::ComputeConstraint(
   // The 'constraint_transform' (submap i <- scan j) is computed from:
   // - a 'high_resolution_point_cloud' in scan j and
   // - the initial guess 'initial_pose' (submap i <- scan j).
+  std::unique_ptr<scan_matching::FastCorrelativeScanMatcher::Result>
+      match_result;
   float score = 0.f;
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc]: Remove these variables? The PR is incomplete: We add the no longer correct values from these unused variables to the histograms.

File: cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc (hunk 2 above)
Comment (gaschler) [reply cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc]: Done

**Comment by gaschler** [comment]
Please take another look.

File: cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher_test.cc [hunk 4]
```diff
@@ -177,10 +180,13 @@ TEST_F(FastCorrelativeScanMatcherTest, CorrectPoseForMatchFullSubmap) {
               transform::IsNearly(result->pose_estimate.cast<float>(), 0.05f))
       << "Actual: " << transform::ToProto(result->pose_estimate).DebugString()
       << "\nExpected: " << transform::ToProto(expected_pose).DebugString();
-  result = fast_correlative_scan_matcher->MatchFullSubmap(
-      Eigen::Quaterniond::Identity(), Eigen::Quaterniond::Identity(),
-      CreateConstantData({Eigen::Vector3f(42.f, 42.f, 42.f)}), kMinScore);
-  EXPECT_THAT(result, testing::IsNull()) << result->low_resolution_score;
+
+  const std::unique_ptr<FastCorrelativeScanMatcher::Result>
```
Comment (wohe) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher_test.cc]: Consistency: Also add `const` above?

Comment (gaschler) [reply cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher_test.cc]: Done

**Comment by gaschler** [comment]
Yeah, I ran a demo bag and confirmed that inter constraints were found.

**Pull request merged**